### PR TITLE
feat(git2go)!: Replace `libgit2` dependency with `go-git`

### LINKED ISSUE #179: Remove libgit2 dependency
This issue tracks removing the [`git2go` dependency](github.com/libgit2/git2go). KraftKit currently makes use of `git2go` which use of an external `libgit2` shared object dependency which is used to calculate the progress of a clone (so when you run `kraft pkg pull` and it resolves to a git repo, the progress bar of the clone is shown).  Whilst useful, this dependency issue causes installation issues and is preventing some new users from using KraftKit.

The use of `git2go` is used only in one location in KraftKit:

https://github.com/unikraft/kraftkit/blob/staging/manifest/pack_pull_git.go#L37

The way to approach a rewrite is to use the almost ubiquitious use of [`go-git/v5`](github.com/go-git/go-git).  [Here's an example of a progress bar based on `v4`](https://gist.github.com/tyru/82cae8bad2b116f442d08eeef456e23e).

### LINKED ISSUE #457: Release binaries are dynamically linked against the outdated libssl1.1
### Description

`make kraft` inside the devenv container produces an executable that is dynamically linked against `libssl1.1`.
By extension this includes **release binaries** (tested with v0.5.1).

This is a problem because the last Ubuntu release that included this version was 20.04 LTS:
- https://packages.ubuntu.com/search?keywords=libssl1.1
- https://packages.ubuntu.com/search?suite=focal&arch=any&searchon=contents&keywords=libssl.so.1.1

```console
$ ldd dist/kraft
        linux-vdso.so.1 (0x00007fffb89b5000)
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f0128905000)
        libssh2.so.1 => /lib/x86_64-linux-gnu/libssh2.so.1 (0x00007f01288c1000)
        libssl.so.1.1 => not found
        libcrypto.so.1.1 => not found
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f01288bc000)
        libresolv.so.2 => /lib/x86_64-linux-gnu/libresolv.so.2 (0x00007f01288a7000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f01286a2000)
        libcrypto.so.3 => /lib/x86_64-linux-gnu/libcrypto.so.3 (0x00007f012824f000)
        libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007f0128231000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f0128914000)
```

`kraft` being a user-oriented program, we must expect people to be running it on modern operating systems.

### PR DESCRIPTION
**Pull request by craciunoiuc**
### Prerequisite checklist

  - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
  - [x] Tested your changes against relevant architectures and platforms;
  - [x] Ran `make fmt` on your commit series before opening this PR;
  - [ ] Updated relevant documentation.

### Description of changes

This PR attempts to remove the `libgit2/git2go` dependency that caused so many protability problems and replace it with a native go library named `go-git/go-git`.
The deciding factor was the inability to easily clone git submodules that are part of a repository.

The new library is not perfect. Currently it lacks printing of progress for the better part of the clone process, which is the download itself and unpacking/checking out.
In order to circumvent this, the last ~80% of the clone process is faked with an exponential function to keep the timer/spinner spinning. The progress bar will block at 98%, as going up to 100% would stop the process. This concerns only `--log-type fancy`.

Once this feature is offered by `go-git`, we will have to go back to the function, as represented by the `FIXME` comment.

To sum it up:

### Advantages
 * Native git support - No more need for CGO
 * 2 MB smaller binary
 * Dependencies handled by Go, and up to date - No need for mimicking kraftkit build system in Related Repositories
 * Overall faster cloning
 * Easier to use with simpler interfaces

### Disadvantages
 * No real progress bar in `fancy` mode

Closes: #179
Closes: #457

### CODE REVIEW

File: manifest/pack_pull_git.go [hunk 1]
```diff
@@ -6,18 +6,119 @@ package manifest
...
+	// Counting objects step
+	countingObjects = iota
+	// Compressing objects step
+	compressingObjects
+	// Downloading objects step
+	downloadingObjects
+)
+
+func (p cloneProgress) Write(b []byte) (int, error) {
+	if p.onProgress != nil {
```
Comment (nderjung) [inline manifest/pack_pull_git.go]:
```diff
- if p.onProgress != nil {
+ if p.onProgress == nil {
+ 	return len(b), nil
+ }
```
Then you can remove an indentation of subsequent lines.

File: .goreleaser-stable.yaml [hunk 2]
```diff
@@ -48,6 +48,4 @@ builds:
       - -X {{ .Env.GOMOD }}/internal/version.buildTime={{ .Date }}
     tags:
       - static
```
Comment (antoineco) [inline .goreleaser-stable.yaml]: Can be removed

File: .goreleaser-staging.yaml [hunk 3]
```diff
@@ -48,6 +48,4 @@ builds:
       - -X {{ .Env.GOMOD }}/internal/version.buildTime={{ .Date }}
     tags:
       - static
```
Comment (antoineco) [inline .goreleaser-staging.yaml]: Can be removed

File: Makefile [hunk 4]
```diff
@@ -85,7 +85,7 @@ endif
 $(addprefix $(.PROXY), $(BIN)): GO_LDFLAGS += -X "$(GOMOD)/internal/version.version=$(VERSION)"
 $(addprefix $(.PROXY), $(BIN)): GO_LDFLAGS += -X "$(GOMOD)/internal/version.commit=$(GIT_SHA)"
 $(addprefix $(.PROXY), $(BIN)): GO_LDFLAGS += -X "$(GOMOD)/internal/version.buildTime=$(shell date)"
-$(addprefix $(.PROXY), $(BIN)): git2go tidy
+$(addprefix $(.PROXY), $(BIN)): tidy
 $(addprefix $(.PROXY), $(BIN)):
 	$(GO) build \
 		-tags static \
```
Comment (antoineco) [inline Makefile]: Can be removed

Comment (craciunoiuc) [reply Makefile]: already is 🤷

Comment (antoineco) [reply Makefile]: I can still see `-tags static` at the latest commit.

Comment (craciunoiuc) [reply Makefile]: Ouch, sorry, I was looking at the modified lines 💫 
image

Comment (craciunoiuc) [reply Makefile]: done

**Comment by antoineco** [comment]
@craciunoiuc related to my previous comments, there is also this leftover in the golangci-lint config:
https://github.com/unikraft/kraftkit/blob/a8ba25d9432931c649fb283b9fbc1942b04f3383/.golangci.yaml#L11-L12

Edit:

There is also one here, but it looked like a mistake to start with because neither this program nor its dependencies use build flags:

https://github.com/unikraft/kraftkit/blob/a8ba25d9432931c649fb283b9fbc1942b04f3383/tools/webinstall/Dockerfile#L22-L23

File: Makefile [hunk 5]
```diff
@@ -128,9 +126,6 @@ test: test-unit test-e2e ## Run all tests.
 .PHONY: test-unit
 test-unit: GOTEST_EXCLUDE := third_party/ test/ hack/ buildenvs/ dist/ docs/
 test-unit: GOTEST_PKGS := $(foreach pkg,$(filter-out $(GOTEST_EXCLUDE),$(wildcard */)),$(pkg)...)
-test-unit: git2go ## Run unit tests.
-	$(GO) run github.com/onsi/ginkgo/v2/ginkgo -v -p -randomize-all --tags static $(GOTEST_PKGS)
-
```
Comment (antoineco) [inline Makefile:133]: Oops. You meant
```diff
-test-unit: git2go ## Run unit tests.
+test-unit: ## Run unit tests.
 	$(GO) run github.com/onsi/ginkgo/v2/ginkgo -v -p -randomize-all --tags static $(GOTEST_PKGS)
```
not
```diff
-test-unit: git2go ## Run unit tests.
-	$(GO) run github.com/onsi/ginkgo/v2/ginkgo -v -p -randomize-all --tags static $(GOTEST_PKGS)
```

File: go.sum [hunk 6]
```diff
@@ -21,11 +21,10 @@ github.com/Microsoft/hcsshim v0.10.0-rc.7/go.mod h1:ILuwjA+kNW+MrN/w5un7n3mTqkws
...
 github.com/OneOfOne/xxhash v1.2.8 h1:31czK/TI9sNkxIKfaUfGlU47BAxQ0ztGgd9vPyqimf8=
-github.com/ProtonMail/go-crypto v0.0.0-20221026131551-cf6655e29de4/go.mod h1:UBYPn8k0D56RtnR8RFQMjmh4KrZzWJ5o7Z9SYjossQ8=
-github.com/ProtonMail/go-crypto v0.0.0-20230214155104-81033d7f4442 h1:OUJ54Fkd+AQXYmr9eOUxZfWNzpK3/e/KD40qa2rKHS4=
-github.com/ProtonMail/go-crypto v0.0.0-20230214155104-81033d7f4442/go.mod h1:I0gYDMZ6Z5GRU7l58bNFSkPTFN6Yl12dsUlAZ8xy98g=
-github.com/acomagu/bufpipe v1.0.3 h1:fxAGrHZTgQ9w5QqVItgzwj235/uYZYgbXitB+dLupOk=
-github.com/acomagu/bufpipe v1.0.3/go.mod h1:mxdxdup/WdsKVreO5GpW4+M/1CE2sMG4jeGJ2sYmHc4=
+github.com/ProtonMail/go-crypto v0.0.0-20230217124315-7d5c6f04bbb8 h1:wPbRQzjjwFc0ih8puEVAOFGELsn1zoIIYdxvML7mDxA=
+github.com/ProtonMail/go-crypto v0.0.0-20230217124315-7d5c6f04bbb8/go.mod h1:I0gYDMZ6Z5GRU7l58bNFSkPTFN6Yl12dsUlAZ8xy98g=
+github.com/acomagu/bufpipe v1.0.4 h1:e3H4WUzM3npvo5uv95QuJM3cQspFNtFBzvJ2oNjKIDQ=
+github.com/acomagu/bufpipe v1.0.4/go.mod h1:mxdxdup/WdsKVreO5GpW4+M/1CE2sMG4jeGJ2sYmHc4=
```
Comment (antoineco) [inline go.sum:28]: This doesn't look right, the go.sum should only lose things, not replace/downgrade.

Comment (craciunoiuc) [reply go.sum:28]: I think that some other things got merged and my staging is no longer up to date, will check now, thanks!

Comment (antoineco) [reply go.sum:28]: (Maybe the key is to leave the upgrade of `go-git/v5` to dependabot?)

File: manifest/pack_pull_git.go [hunk 7]
```diff
@@ -6,18 +6,99 @@ package manifest
...
+	var scale float64
+
+	// Trim prefix
+	if strings.HasPrefix(string(b), "Counting objects") {
+		prog = strings.TrimPrefix(string(b), "Counting objects: ")
+		scale = 0.1
+	} else if strings.HasPrefix(string(b), "Compressing objects") {
+		prog = strings.TrimPrefix(string(b), "Compressing objects: ")
+		scale = 0.2
+	} else if strings.HasPrefix(string(b), "Total ") {
```
Comment (antoineco) [inline manifest/pack_pull_git.go]: nit(style): a good candidate for a `switch`

```suggestion
	switch {
	case strings.HasPrefix(string(b), "Counting objects"):
		prog = strings.TrimPrefix(string(b), "Counting objects: ")
		scale = 0.1
	case strings.HasPrefix(string(b), "Compressing objects"):
		prog = strings.TrimPrefix(string(b), "Compressing objects: ")
		scale = 0.2
	case strings.HasPrefix(string(b), "Total "):
```

File: manifest/pack_pull_git.go [hunk 8]
```diff
@@ -6,18 +6,99 @@ package manifest
...
+	githttp "github.com/go-git/go-git/v5/plumbing/transport/http"
+	gitssh "github.com/go-git/go-git/v5/plumbing/transport/ssh"
 
 	"kraftkit.sh/log"
 	"kraftkit.sh/pack"
 	"kraftkit.sh/unikraft"
 )
 
+type cloneProgress struct {
+	io.Writer
```
Comment (antoineco) [inline manifest/pack_pull_git.go]: This seems unused, why is it embedded?

File: manifest/pack_pull_git.go [hunk 9]
```diff
@@ -36,72 +117,62 @@ func pullGit(ctx context.Context, manifest *Manifest, opts ...pack.PullOption) e
...
-
-		copts.FetchOptions.RemoteCallbacks.CredentialsCallback = func(_, usernameFromUrl string, allowedTypes libgit.CredentialType) (*libgit.Credential, error) {
-			if allowedTypes == libgit.CredentialTypeUsername && len(usernameFromUrl) == 0 {
-				return nil, fmt.Errorf("username required in SSH URL")
-			}
-
-			return libgit.NewCredentialSSHKeyFromAgent(usernameFromUrl)
+		copts.Auth, err = gitssh.NewSSHAgentAuth("git")
+		if err != nil {
+			return fmt.Errorf("could not create SSH agent auth: %v", err)
```
Comment (antoineco) [inline manifest/pack_pull_git.go]: By wrapping errors we can later discriminate them with `errors.Is`, `errors.Has` and `errors.Unwrap`. In some cases this can be very beneficial, so it's a good practive to always use `%w` with `Errorf`.

```suggestion
			return fmt.Errorf("could not create SSH agent auth: %w", err)
```

File: manifest/pack_pull_git.go [hunk 10]
```diff
@@ -36,72 +117,62 @@ func pullGit(ctx context.Context, manifest *Manifest, opts ...pack.PullOption) e
...
-
-			if auth, ok := manifest.Auths()[u.Host]; ok {
-				if len(auth.User) > 0 && len(auth.Token) > 0 {
-					return libgit.NewCredentialUserpassPlaintext(auth.User, auth.Token)
+		if !strings.HasPrefix(path, "https://") {
+			path = "https://" + path
+		}
+		u, err := url.Parse(path)
+		if err != nil {
+			return fmt.Errorf("could not parse URL: %v", err)
```
Comment (antoineco) [inline manifest/pack_pull_git.go]: For the same reason as my previous comment.

```suggestion
			return fmt.Errorf("could not parse URL: %w", err)
```

File: manifest/pack_pull_git.go [hunk 11]
```diff
@@ -6,18 +6,99 @@ package manifest
...
 	"kraftkit.sh/unikraft"
 )
 
+type cloneProgress struct {
+	io.Writer
+
+	once           sync.Once
+	onProgress     func(progress float64)
+	completeParent chan bool
+	completeWorker chan bool
```
Comment (antoineco) [inline manifest/pack_pull_git.go]: The boolean value isn't used, so we may as well use zero values (an empty struct is usually considered idiomatic for that).

```suggestion
	completeParent chan struct{}
	completeWorker chan struct{}
```

File: manifest/pack_pull_git.go (hunk 8 above)
Comment (craciunoiuc) [reply manifest/pack_pull_git.go]: The writer is needed by the Progress function. `go-git` passes data to us with the writer

Comment (craciunoiuc) [reply manifest/pack_pull_git.go]: works without it, nvm 😅

File: Makefile [hunk 12]
```diff
@@ -128,9 +126,9 @@ test: test-unit test-e2e ## Run all tests.
 .PHONY: test-unit
 test-unit: GOTEST_EXCLUDE := third_party/ test/ hack/ buildenvs/ dist/ docs/
 test-unit: GOTEST_PKGS := $(foreach pkg,$(filter-out $(GOTEST_EXCLUDE),$(wildcard */)),$(pkg)...)
-test-unit: git2go ## Run unit tests.
-	$(GO) run github.com/onsi/ginkgo/v2/ginkgo -v -p -randomize-all --tags static $(GOTEST_PKGS)
 
+test-unit: ## Run unit tests.
+	$(GO) run github.com/onsi/ginkgo/v2/ginkgo -v -p -randomize-all $(GOTEST_PKGS)
```
Comment (antoineco) [inline Makefile]: This new line here doesn't look intentional.

**Review by antoineco — COMMENTED** [review]
Apart from my last comment, and the dependencies versions which I think we should avoid touching in the PR, the code looks fine after a cursory look (not tested myself!).

File: Makefile (hunk 12 above)
Comment (craciunoiuc) [reply Makefile]: well you asked me to add it back :D

File: go.sum (hunk 6 above)
Comment (craciunoiuc) [reply go.sum:28]: It replaces/upgrades most of them. There is just 1 downgrade. I think this happens because we are replacing it with a very similar library, and they are pinned at different versions

File: Makefile (hunk 12 above)
Comment (antoineco) [reply Makefile]: I'm talking about the new line above. It isn't there on `main`:

https://github.com/unikraft/kraftkit/blob/2609f3a0b160a555525fc5d76f88fa8cec7afa6e/Makefile#L128-L132

Comment (craciunoiuc) [reply Makefile]: whoops 😊

**Review by antoineco — APPROVED** [review]

**Comment by nderjung** [comment]
Hi. I tried this out locally, this works for me. Good to go after the `go.{mod,sum}` are rebased. Thanks!

**Review by nderjung — APPROVED** [review]
😭 Thank you
